File: Study5 - Tree/Week27/Day187/bj15900_hsk.js
const fs = require('fs');
let stdin = (
  process.platform === 'linux'
    ? fs.readFileSync('/dev/stdin').toString().trim()
    : `2
2 1`
).split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const calcDepthFromRoot = (nodeAndDepthArr) => {
  while (nodeAndDepthArr.length > 0) {
    let [nodeNum, depth] = nodeAndDepthArr.pop();
    visitedNode[nodeNum] = true;

    let connectedNodeList = tree[nodeNum];
    if (connectedNodeList.length === 1 && nodeNum !== 1) {
      depthFromRootList.push(depth);
      continue;
    }

    connectedNodeList.forEach((node) => {
      if (!visitedNode[node]) nodeAndDepthArr.push([node, depth + 1]);
    });
  }
};

const sumDepthFromRoot = (depthArr) => {
  return depthArr.reduce((sum, curr) => sum + curr, 0);
};

const isWinGame = (num) => {
  if (num % 2 !== 0) return true;
  return false;
};

const N = parseInt(input());
const tree = Array.from(Array(N + 1), () => Array());
const visitedNode = new Array(N + 1).fill(false);
let depthFromRootList = [];
let nodeAndDepthArr = [[1, 0]];

for (let i = 1; i < N; i++) {
  const [nodeA, nodeB] = input()
    .split(' ')
    .map((value) => parseInt(value));
  tree[nodeA].push(nodeB);
  tree[nodeB].push(nodeA);
}

calcDepthFromRoot(nodeAndDepthArr);

let sumDepth = sumDepthFromRoot(depthFromRootList);
console.log(isWinGame(sumDepth) ? 'Yes' : 'No');

File: Study5 - Tree/Week27/Day187/bj15900_hsk.py
# nodejs 에서 시간초과와 재귀오류를 해결하지 못하여 python으로 풀었음
import sys


N = int(input())
tree = [[] for _ in range(N + 1)]
visitedNode = [0] * (N + 1)
depthFromRootList = []
nodeAndDepthArr = [[1, 0]]

for i in range(N - 1):
    nodeA, nodeB = map(int, sys.stdin.readline().rstrip().split())
    tree[nodeA].append(nodeB)
    tree[nodeB].append(nodeA)

while nodeAndDepthArr:
    nodeNum, depth = nodeAndDepthArr.pop()
    visitedNode[nodeNum] = 1

    connectedNodeList = tree[nodeNum]
    if nodeNum != 1 and len(connectedNodeList) == 1:
        depthFromRootList.append(depth)
        continue

    for node in connectedNodeList:
        if visitedNode[node] == 0:
            nodeAndDepthArr.append([node, depth + 1])

## 모두 구하는 방법으로도 가능하군요! 다른 방법도 알아갑니다.
sumDepth = sum(depthFromRootList)

if sumDepth % 2:
    print('Yes')
else:
    print('No')
